drv/cosmo-hf/src/hf.rs

```diff
@@ -53,6 +71,134 @@ impl ServerImpl {
+            if data.is_valid() && best.map(|b| data > b).unwrap_or(true) {
+                best = Some(data);
+            }
+            if empty_slot.is_none()
+                && data.as_bytes().iter().all(|b| *b == 0xFF)
+            {
+                empty_slot = Some(addr);
+            }
+        }
+        (best, empty_slot)
+    }
+
+    fn get_persistent_data(&mut self) -> Result<HfPersistentData, HfError> {
+        let out = self.get_raw_persistent_data()?;
+        Ok(HfPersistentData {
+            dev_select: match out.dev_select {
+                0 => HfDevSelect::Flash0,
+                _ => HfDevSelect::Flash1,
+            },
+        })
+    }
+
+    /// Reads persistent data from both slots, returning the newest
+    fn get_raw_persistent_data(
+        &mut self,
+    ) -> Result<HfRawPersistentData, HfError> {
+        let prev_slot = self.dev;
+
+        // Look at the inactive slot first
+        self.dev = !prev_slot;
+        let (a, _) = self.persistent_data_scan();
+
+        // Then switch back to our current slot, so that the resulting state
+        // is unchanged.
+        self.dev = prev_slot;
+        let (b, _) = self.persistent_data_scan();
+
+        match (a, b) {
+            (Some(a), Some(b)) => Some(a.max(b)),
+            (Some(_), None) => a,
+            (None, Some(_)) => b,
+            (None, None) => None,
+        }
+        .ok_or(HfError::NoPersistentData)
+    }
+
+    /// Writes raw persistent data to the given address on the
+    /// currently-selected virtual flash device.
+    ///
+    /// If `addr` is `None`, then we're out of available space; erase all of
+    /// sector 0 and write to address 0 upon success.
+    fn write_raw_persistent_data_to_addr(
+        &mut self,
+        addr: Option<u32>,
+        raw_data: &HfRawPersistentData,
+    ) {
+        let addr = match addr {
+            Some(a) => self.flash_addr(a),
+            None => {
+                let addr = self.flash_addr(0);
+                self.drv.flash_sector_erase(addr);
+                addr
+            }
+        };
+        // flash_write is infallible when given a slice
+        self.drv
+            .flash_write(addr, &mut raw_data.as_bytes())
+            .unwrap_lite()
+    }
+
+    /// Ensures that the persistent data is consistent between the virtual devs
+    fn ensure_persistent_data_is_redundant(&mut self) {
+        // Load the current state of persistent data from flash
+        let prev_slot = self.dev;
+
+        self.dev = !prev_slot;
+        let (a_data, a_next) = self.persistent_data_scan();
+
+        self.dev = prev_slot;
+        let (b_data, b_next) = self.persistent_data_scan();
+
+        match (a_data, b_data) {
+            (Some(a), Some(b)) => {
+                match a.cmp(&b) {
+                    core::cmp::Ordering::Less => {
+                        self.dev = !prev_slot;
+                        self.write_raw_persistent_data_to_addr(a_next, &b);
```

Comment:
similarly to my comments on the calls to `persistent_data_scan`, what do you think about just passing the `HfDevSelect` to this function rather than mutating state we have to then remember to go back and un-mutate?